### PR TITLE
CLN/STY: pandas/_libs/internals.pyx

### PR DESCRIPTION
**Pull request by ShaharNaveh**
- [ ] closes #xxxx
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/_libs/internals.pyx [hunk 1]
```diff
@@ -370,22 +369,16 @@ def get_blkno_indexers(int64_t[:] blknos, bint group=True):
     # There's blkno in this function's name because it's used in block &
     # blockno handling.
     cdef:
-        int64_t cur_blkno
-        Py_ssize_t i, start, stop, n, diff
-
+        Py_ssize_t i, start = 0, stop, n = blknos.shape[0], diff, tot_len
```
Comment (MomIsBestFriend) [inline pandas/_libs/internals.pyx]: NOTE, I am adding here a variable definition (```tot_len```)

File: pandas/_libs/internals.pyx [hunk 2]
```diff
@@ -52,81 +51,86 @@ cdef class BlockPlacement:
     def __str__(self) -> str:
         cdef:
             slice s = self._ensure_has_slice()
-        if s is not None:
-            v = self._as_slice
-        else:
-            v = self._as_array
 
-        return f'{type(self).__name__}({v})'
+        v = self._as_slice if s is not None else self._as_array
```
Comment (jorisvandenbossche) [inline pandas/_libs/internals.pyx]: Personally, I don't find this necessarily more readable ..

File: pandas/_libs/internals.pyx [hunk 3]
```diff
@@ -52,81 +51,86 @@ cdef class BlockPlacement:
...
         return str(self)
 
     def __len__(self) -> int:
         cdef:
             slice s = self._ensure_has_slice()
+
         if s is not None:
             return slice_len(s)
-        else:
-            return len(self._as_array)
```
Comment (jorisvandenbossche) [inline pandas/_libs/internals.pyx]: Why this change? I personally find an if/else pattern very clear

File: pandas/_libs/internals.pyx (hunk 1 above)
Comment (jorisvandenbossche) [reply pandas/_libs/internals.pyx]: I would put this on multiple lines, this becomes difficult to read IMO

File: pandas/_libs/internals.pyx [hunk 4]
```diff
@@ -159,14 +164,10 @@ cdef class BlockPlacement:
...
-            if ((step > 0 and start < 0) or
-                    (step < 0 and stop < step)):
+            if (step > 0 and start < 0) or (step < 0 and stop < step):
                 raise ValueError("iadd causes length change")
 
-            if stop < 0:
-                val = slice(start, None, step)
-            else:
-                val = slice(start, stop, step)
+            val = slice(start, None, step) if stop < 0 else slice(start, stop, step)
```
Comment (jorisvandenbossche) [inline pandas/_libs/internals.pyx]: same here

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]
Since code style is subjective (when going beyond PEP8 / black), you also get some subjective comments from me ;)

File: pandas/_libs/internals.pyx [hunk 5]
```diff
@@ -74,59 +75,65 @@ cdef class BlockPlacement:
...
             slice s = self._ensure_has_slice()
             Py_ssize_t start, stop, step, _
+
         if s is not None:
             start, stop, step, _ = slice_get_indices_ex(s)
             return iter(range(start, stop, step))
-        else:
-            return iter(self._as_array)
+
+        return iter(self._as_array)
```
Comment (MomIsBestFriend) [inline pandas/_libs/internals.pyx]: @jorisvandenbossche This is a very similar change to the change that was mentioned [here](https://github.com/pandas-dev/pandas/pull/32801#discussion_r394308161), would you like me to revert that as well?

File: pandas/_libs/internals.pyx [hunk 6]
```diff
@@ -74,59 +75,65 @@ cdef class BlockPlacement:
...
+
     @property
     def indexer(self):
         cdef:
             slice s = self._ensure_has_slice()
+
         if s is not None:
             return s
-        else:
-            return self._as_array
```
Comment (MomIsBestFriend) [inline pandas/_libs/internals.pyx]: @jorisvandenbossche This is a very similar change to the change that was mentioned [here](https://github.com/pandas-dev/pandas/pull/32801#discussion_r394308161), would you like me to revert that as well?

File: pandas/_libs/internals.pyx [hunk 7]
```diff
@@ -74,59 +75,65 @@ cdef class BlockPlacement:
...
 
     def __getitem__(self, loc):
         cdef:
             slice s = self._ensure_has_slice()
-        if s is not None:
-            val = slice_getitem(s, loc)
-        else:
-            val = self._as_array[loc]
+
+        val = slice_getitem(s, loc) if s is not None else self._as_array[loc]
```
Comment (MomIsBestFriend) [inline pandas/_libs/internals.pyx]: @jorisvandenbossche This is a very similar change to the change that was mentioned [here](https://github.com/pandas-dev/pandas/pull/32801#discussion_r394307873), would you like me to revert that as well?

Comment (jbrockmendel) [reply pandas/_libs/internals.pyx]: this is different from the other places where you left this comment, and this is the only one i would ask you to revert.  The way it is currently written, I can look in coverage output to see if both branches are reached; I can't with this PR's version.

File: pandas/_libs/internals.pyx [hunk 8]
```diff
@@ -74,55 +75,64 @@ cdef class BlockPlacement:
...
         cdef:
             Py_ssize_t start, stop, end, _
+
         if not self._has_array:
             start, stop, step, _ = slice_get_indices_ex(self._as_slice)
-            # NOTE: this is the C-optimized equivalent of
-            #  np.arange(start, stop, step, dtype=np.int64)
+            # NOTE:
+            # this is the C-optimized equivalent of
+            # `np.arange(start, stop, step, dtype=np.int64)`
```
Comment (jbrockmendel) [inline pandas/_libs/internals.pyx]: not a deal-breaker but i dont see how this is an improvement.  especially if the comment is going to be repeated in multiple places id rather it be 2 lines than 3

File: pandas/_libs/internals.pyx [hunk 9]
```diff
@@ -370,22 +378,17 @@ def get_blkno_indexers(int64_t[:] blknos, bint group=True):
...
-    n = blknos.shape[0]
+        int64_t cur_blkno = blknos[start]
 
     if n == 0:
         return
 
-    start = 0
-    cur_blkno = blknos[start]
-
-    if group is False:
```
Comment (jbrockmendel) [inline pandas/_libs/internals.pyx]: `if group is False` should be marginally faster than `if not group`

File: pandas/_libs/internals.pyx [hunk 10]
```diff
@@ -370,22 +378,17 @@ def get_blkno_indexers(int64_t[:] blknos, bint group=True):
...
         int64_t[:] res_view
-
-    n = blknos.shape[0]
+        int64_t cur_blkno = blknos[start]
 
     if n == 0:
         return
 
-    start = 0
-    cur_blkno = blknos[start]
```
Comment (jbrockmendel) [inline pandas/_libs/internals.pyx]: doing this here instead of above means we can avoid the assignment in the n==0 case

File: pandas/_libs/internals.pyx (hunk 9 above)
Comment (MomIsBestFriend) [reply pandas/_libs/internals.pyx]: From my benchmarking it's actually the other way around:

```
In [1]: foo = True                                                                                            

In [2]: %timeit foo is False                                                                                  
22.9 ns ± 0.157 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [3]: %timeit foo is False                                                                                  
23 ns ± 0.0388 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [4]: %timeit foo is False                                                                                  
22.7 ns ± 0.194 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [5]: %timeit not foo                                                                                       
20.7 ns ± 0.0589 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [6]: %timeit not foo                                                                                       
21.1 ns ± 0.0636 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [7]: %timeit not foo                                                                                       
20.6 ns ± 0.0497 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

---

```

In [1]: bar = False                                                                                           

In [2]: %timeit bar is False                                                                                  
28.5 ns ± 0.077 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [3]: %timeit bar is False                                                                                  
28.6 ns ± 0.0666 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [4]: %timeit bar is False                                                                                  
29.5 ns ± 0.203 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [5]: %timeit not bar                                                                                       
26.1 ns ± 0.217 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [6]: %timeit not bar                                                                                       
25.8 ns ± 0.0231 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [7]: %timeit not bar                                                                                       
26.7 ns ± 0.0831 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

Comment (jbrockmendel) [reply pandas/_libs/internals.pyx]: look at the generated C code.  The "is" should be a pointer comparison, whereas the other one should have to do some more work.

**Review by jreback — COMMENTED** [review]
looks fine. this is mostly moving things around, but ok.

File: pandas/_libs/internals.pyx (hunk 6 above)
Comment (jorisvandenbossche) [reply pandas/_libs/internals.pyx]: To be clear, my opinion is only a single one, but yes, I would personally revert this as well

File: pandas/_libs/internals.pyx (hunk 5 above)
Comment (jorisvandenbossche) [reply pandas/_libs/internals.pyx]: same here (see comment below)

File: pandas/_libs/internals.pyx [hunk 11]
```diff
@@ -74,55 +75,63 @@ cdef class BlockPlacement:
...
         cdef:
             slice s = self._ensure_has_slice()
-        if s is None:
-            raise TypeError('Not slice-like')
-        else:
+
+        if s is not None:
             return s
 
+        raise TypeError("Not slice-like")
```
Comment (jorisvandenbossche) [inline pandas/_libs/internals.pyx]: Also this change I personally don't find an improvement in code flow

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jreback** [comment]
thanks @MomIsBestFriend
